Automattic/newspack-plugin
# feat(reader-activation): account link and auth form

### All Submissions:






### Changes proposed in this Pull Request:

Implements a multi-strategy sign-in form and an always-visible responsive account/sign-in link.

**The form only signs in users identified as readers.**

If a reader is not fully authenticated, clicking on the link triggers the auth form. If the reader is authenticated, the link redirects to WC's "My Account" page.

The form button uses the theme's `primary_color` or Newspack's `#36f` as a fallback.

If a reader is pre-authenticated (auth intention), the form email input is automatically filled with the stored email.

## Desktop

The account link is displayed hooking into the Newspack theme's `tertiary-menu` location. In case the location doesn't have a menu assigned to it, it'll hook as a custom HTML using `fallback_cb`.

<img width="850" alt="image" src="https://user-images.githubusercontent.com/820752/178543762-ec4f50d3-1fa1-4733-810f-59e7bb8392d0.png">

<img width="615" alt="image" src="https://user-images.githubusercontent.com/820752/178543853-a0bda01b-2eef-4f1e-97b1-de6ff884cb39.png">

<img width="613" alt="image" src="https://user-images.githubusercontent.com/820752/178543893-04eab2d2-9fef-4105-a464-0790afc2deda.png">

<img width="596" alt="image" src="https://user-images.githubusercontent.com/820752/178543943-d12a42fb-2e69-45a0-bfe2-f945913ff3aa.png">

## Mobile

The account link is displayed hooking into the Newspack theme's [`newspack_header_after_mobile_toggle` custom action](https://github.com/Automattic/newspack-theme/pull/1875).

<img width="359" alt="image" src="https://user-images.githubusercontent.com/820752/178544087-81ddcb3e-0b9b-42d8-ac5f-d464fbf8fc11.png">

<img width="359" alt="image" src="https://user-images.githubusercontent.com/820752/178544120-8fd24192-19bd-4c49-a385-90691ef8ff72.png">

<img width="361" alt="image" src="https://user-images.githubusercontent.com/820752/178544188-35e54c54-1c06-4dca-b641-8de46bd8d95c.png">

Closes #1746 
Closes #1749
Closes #1744

## Form next steps

 - Third-party auth
 - Newsletters integration for registration
 - Reader Activation settings management on the Engagement wizard for:
   - Toggling account link visibility
   - Configuring default messages (privacy policy)
  


### How to test the changes in this Pull Request:

1. Check out this branch and https://github.com/Automattic/newspack-theme/pull/1875
2. Make sure you have reader activation enabled and AMP plugin off (we'll later test AMP Plus)
3. Visit the homepage anonymously and confirm you see the "Sign In" link correctly placed
4. Test authenticating your admin account and confirm it displays an error
5. Test authenticating a reader through auth link and password and confirm the account link UI is updated
6. Click on the account link after authenticating and confirm you are redirected to the "My Account" page
7. Turn AMP on with AMP Plus
8. Repeat the previous steps and confirm the form works as expected
9. Visit the site anonymously on a mobile viewport
10. Test authentication and confirm the form behaves as expected

### Other information:
